rclcpp/test/rclcpp/topic_statistics/test_subscription_topic_statistics.cpp

```diff
@@ -188,6 +188,12 @@ class SubscriberWithTopicStatistics : public rclcpp::Node
     auto options = rclcpp::SubscriptionOptions();
     options.topic_stats_options.state = rclcpp::TopicStatisticsState::Enable;
     options.topic_stats_options.publish_period = publish_period;
+    // There is a possibility that age and period messages are published almost at the
+    // same time, and if the message order is inverted when storing DDS HistoryCache
+    // 1st message could be lost with system default QoS keep_last.
+    // We explicitly set the QoS with keep_all for the test, so that both messages
+    // will be stored in the HistoryCache to be sent out.
+    options.topic_stats_options.qos.keep_all();
```

Comment:
Do we really want to `keep_all`?  This can effectively use up all memory.  What about `keep_last`, but with a larger `depth`?

Reply:
i think `keep_last` does not work with connextdds, current Statistics Publisher QoS is system default, so connextdds sets it to `keep_last` with some depth. i think the root cause is that message can be inverted order to the DataWriter HistoryCache, in that case, `keep_last` does not work... i am not sure why only connextdds has this situation 100%. (cyclonedds `publish` is sync manner, so this call guarantees to put the message in the DDS HistoryCache, that what i believe. but Fast-DDS uses async `publish` in default, but never met this problem.)

Reply:
@clalancette i think i was wrong. i tested with `keep_last(10)` that seems to work without any failure with connextdds. i was misunderstanding that `keep_last` only takes the recent messages in the queue no matter what the depth is. since the test is passing with `keep_last(10)`, connextdds default depth was just not enough? probably depth is 1?

@fgallegosalido can you confirm my thought above?

Reply:
So, actually, it is interesting.

As you say, by default topic_stats_options is SystemDefaultQoS.  In turn, that is initialized from `rmw_qos_profile_system_default`.  But if you look closely at https://github.com/ros2/rmw/blob/33118c9d4dc2adec838962554f0e09ab5c15d1e0/rmw/include/rmw/qos_profiles.h#L93 , you can see that it is set to RMW_QOS_POLICY_DEPTH_SYSTEM_DEFAULT.  And that particular setting means that the RMW is free to choose what to use as the default.  Looking at our RMWs:

* [rmw_cyclonedds_cpp](https://github.com/ros2/rmw_cyclonedds/blob/0c2904d725dade83cc0b1ffde931b46f9d28269f/rmw_cyclonedds_cpp/src/rmw_node.cpp#L2163) chooses 1 as the default
* rmw_fastrtps_cpp is a little harder to tell, but I believe that if nothing is specified they rely on the [default from Fast-DDS itself, which is also 1](https://github.com/eProsima/Fast-DDS/blob/a8396fa4c56a5499082fd5fdb5dc2f0ddeb1f659/include/fastdds/dds/core/policy/QosPolicies.hpp#L1679)
* [rmw_connextdds](https://github.com/ros2/rmw_connextdds/blob/92f0925e90094519351cfe7db05d8f282bfe85d3/rmw_connextdds_common/src/common/rmw_impl.cpp#L330) also chooses 1 as the default.

So that explains why using a depth of 10 helps.  I don't know why the failure only occurs with Connext, but it does seem like there is potential for it to be a problem for all RMWs.

I'd still like to hear back from @fgallegosalido, but I think we actually should probably explicitly set this to `keep_last` and a depth of 10 in https://github.com/ros2/rclcpp/blob/0be8aa013e4ed94bcab88527887e85130816d256/rclcpp/include/rclcpp/subscription_options.hpp#L83 , because real users (not just tests) could run into this issue as well.

Reply:
@clalancette appreciate your information! thanks.

> cyclonedds publish is sync manner, so this call guarantees to put the message in the DDS HistoryCache

cyclonedds is sync publish, so i think this should be fine. but i cannot explain Fast-DDS case, probably internal pre-cache queue for async publish would be different behavior.

Reply:
> but Fast-DDS uses async `publish` in default

@fujitatomoya Just FYI, Fast DDS uses **sync** publishing by default

Reply:
> @clalancette i think i was wrong. i tested with `keep_last(10)` that seems to work without any failure with connextdds. i was misunderstanding that `keep_last` only takes the recent messages in the queue no matter what the depth is. since the test is passing with `keep_last(10)`, connextdds default depth was just not enough? probably depth is 1?
> 
> @fgallegosalido can you confirm my thought above?

Connext DDS also uses depth 1 by default. As @clalancette mentioned, setting the history depth to 10 should fix most inestabilities.

Reply:
> Fast DDS uses sync publishing by default

that is the reason, https://github.com/ros2/rmw_fastrtps?tab=readme-ov-file#change-publication-mode thanks for pointing that out.